### ISSUE TITLE
prov/gni __gnix_nic_prog_thread_fn() GNI_CqVectorMonitor does not process CQs for a which value of 0

### ISSUE THREAD

**Issue by tonyzinger** [ofiwg/libfabric#3660]
The GNI_CqVectorMonitor() call can return a GNI_RC_SUCCESS status with a which value of 0.
This can mean 2 things.  The first meaning is that CQ associated with the index, which value, of 0 has a CQ event available.  The second meaning is that there is a DLA CQ event that needs to be processed.  In the second case a call must be made to GNI_CqGetEvent() to process the DLA CQ.

From the function documentation for GNI_CqVecectorMonitor:
 * There is a case where the user will need to process the DLA CQ events
 * before they will receive their CQ events.  This is done by returning
 * a GNI_RC_SUCCESS status and with the which value set to 0.  This will
 * trick the user into calling GNI_CqGetEvent() with a valid CQ to process
 * the DLA CQ events. For this case, there probably will not be any user
 * CQ events available.

**Comment by hppritcha** [comment]
Are you seeing indications that the  CQE's are getting dequeued from the tx_cq_blk?  Looking at line
609 to 613 of gnix_nic.c it looks like CQEs are being dequeued.

```
¬       if (nic->tx_cq_blk && nic->tx_cq_blk != nic->tx_cq) {
¬       ¬       ret =  __nic_tx_progress(nic, nic->tx_cq_blk);
¬       ¬       if (OFI_UNLIKELY(ret != FI_SUCCESS))
¬       ¬       ¬       return ret;
¬       }
```

**Comment by tonyzinger** [comment]
Here is the debug output from libfabric, when it hangs trying to do a fi_cntr_open()
2: libfabric:gni:cq:gnix_cntr_open():433<trace> [15239:1]
2: libfabric:gni:cq:__verify_cntr_attr():65<trace> [15239:1]
2: libfabric:gni:core:_gnix_ref_init():254<debug> [15239:1] 0x650298 refs 1
2: libfabric:gni:core:gnix_cntr_open():463<debug> [15239:1] 0x605690 refs 5
2: libfabric:gni:eq:gnix_cntr_set_wait():96<trace> [15239:1]
2: libfabric:gni:eq:gnix_wait_open():530<trace> [15239:1]
2: libfabric:gni:eq:gnix_verify_wait_attr():361<trace> [15239:1]
2: libfabric:gni:eq:gnix_init_wait_obj():381<trace> [15239:1]
2: libfabric:gni:core:gnix_wait_open():558<debug> [15239:1] 0x605598 refs 3
2: libfabric:gni:ep_ctrl:_gnix_dgram_prog_thread_fn():79<trace> [15239:5]
2: libfabric:gni:ep_ctrl:_gnix_dgram_poll():436<trace> [15239:5]
2: libfabric:gni:ep_ctrl:__gnix_nic_prog_thread_fn():92<trace> [15239:6]

**Comment by hppritcha** [comment]
I think I'll need a reproducer to try and figure out what's going wrong here.

**Comment by tonyzinger** [comment]
There is a README file and a test case on jupiter in the directory /home/users/ajz/issue_3660

**Comment by hppritcha** [comment]
thanks I"ll take a look once I get my secureid resync'd.  my cray accounts were disabled prior to the holiday shutdown.

**Comment by tonyzinger** [comment]
I have created a simpler test case.  This test case creates a libfabric environment, tears it down and then creates a second libfabric environment.

This test case executes successfully using the sockets provider.

Here is the output from my test case using the gni provider without libfabric debug:
srun -l --exclusive --ntasks=1 --ntasks-per-node=1 -t 2 /home/users/ajz/issue_3660/hang_test -P gni -I 2
0: [nid00020, Rank:  0] DEBUG: fi_allocinfo() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_getinfo() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_fabric() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_domain() succeeded
0: [nid00020, Rank:  0] DEBUG: ltu_fabric_init() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_endpoint() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_cq_open() src_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_ep_bind() src_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_cq_open() rcv_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_ep_bind() rcv_cq succeeded
0: [nid00020, Rank:  0] DEBUG: ltu_ep_init: calling fi_cntr_open
0: [nid00020, Rank:  0] DEBUG: fi_cntr_open() rcv_cntr succeeded
0: [nid00020, Rank:  0] DEBUG: fi_ep_bind() rcv_cntr succeeded
0: [nid00020, Rank:  0] DEBUG: ltu_ep_init() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_av_open() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_ep_bind() addr_vec succeeded
0: [nid00020, Rank:  0] DEBUG: fi_getname() local succeeded
0: [nid00020, Rank:  0] DEBUG: fi_av_insert() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_enable() succeeded
0: [nid00020, Rank:  0] DEBUG: ltu_av_init() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_close() ep succeeded
0: [nid00020, Rank:  0] DEBUG: fi_close() av succeeded
0: [nid00020, Rank:  0] DEBUG: fi_close() rcv_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_close() src_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_close() domain succeeded
0: [nid00020, Rank:  0] DEBUG: fi_close() fabric succeeded
0: [nid00020, Rank:  0] DEBUG: ltu_fabric_close() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_allocinfo() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_getinfo() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_fabric() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_domain() succeeded
0: [nid00020, Rank:  0] DEBUG: ltu_fabric_init() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_endpoint() succeeded
0: [nid00020, Rank:  0] DEBUG: fi_cq_open() src_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_ep_bind() src_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_cq_open() rcv_cq succeeded
0: [nid00020, Rank:  0] DEBUG: fi_ep_bind() rcv_cq succeeded
0: [nid00020, Rank:  0] DEBUG: ltu_ep_init: calling fi_cntr_open
*** the test hangs at this point

Here is the libfabric debug information, starting at the "ltu_ep_init: calling fi_cntr_open()":
0: [nid00020, Rank:  0] DEBUG: ltu_ep_init: calling fi_cntr_open
0: libfabric:gni:ep_ctrl:gnix_ep_bind():1809<trace> [44217:1]
0: libfabric:gni:core:gnix_ep_bind():1901<debug> [44217:1] 0x6500f8 refs 2
0: libfabric:gni:cq:gnix_cntr_open():433<trace> [44217:1]
0: libfabric:gni:cq:__verify_cntr_attr():65<trace> [44217:1]
0: libfabric:gni:core:_gnix_ref_init():254<debug> [44217:1] 0x634198 refs 1
0: libfabric:gni:core:gnix_cntr_open():463<debug> [44217:1] 0x60a910 refs 5
0: libfabric:gni:eq:gnix_cntr_set_wait():96<trace> [44217:1]
0: libfabric:gni:eq:gnix_wait_open():530<trace> [44217:1]
0: libfabric:gni:eq:gnix_verify_wait_attr():361<trace> [44217:1]
0: libfabric:gni:eq:gnix_init_wait_obj():381<trace> [44217:1]
0: libfabric:gni:core:gnix_wait_open():558<debug> [44217:1] 0x60a7a8 refs 3
0: libfabric:gni:ep_ctrl:_gnix_dgram_poll():436<trace> [44217:6]

**Comment by hppritcha** [comment]
@tonyzinger I still don't have access to jupiter.  Could you email me the test or put in a gist where I can access?  SecureID died and don't have a replacement yet.

**Comment by tonyzinger** [comment]
Howard,  I am creating a new issue for the fi_cntr_open() hang problem.
I inadvertently combined 2 problems in this issue.

The new issue that I created is: #3777

**Comment by hppritcha** [comment]
@tonyzinger fix for the ```fi_cntr_open``` has been merged in to master.  could you check this test now?

**Comment by tonyzinger** [comment]
My test case for the fi_cntr_open hang when creating, destroying and then re-creating a libfabric environment works successfully.

Thanks, Howard.

**Comment by hppritcha** [comment]
So can we close this issue as well?  Or was there another problem that was contributing to the failure of the test at /home/users/ajz/issue_3660 on jupiter?

**Comment by tonyzinger** [comment]
When I was trying to track down a hang problem in one of my tests, I must have turned on debugging and noticed the GNI_CqVectorMonitor call.  Then I noticed that CQ events were only being processed for a which value of 1.  I did not follow the code sequence after this point.  So if _gnix_nic_progress() processes the CQ that is associated with which of 0.  Then there probably is not a problem.

The thing that I was concerned about was that when GNI_CqVectorMonitor() returns a which value of 0, this does not necessarily mean that the CQ associated with a which of 0 has an event waiting.  This can also mean that the DLA CQs need to be processed, which can release DLA credits.

This is also the reason why I created issue #3777, the other day to split up these 2 problems/concerns.
The GNI_CqVectorMonitor processing the which value of 0 (this issue) and the fi_cntr_open hang (issue #3777).

Therefore, if the CQ associated with a which value of 0 from GNI_CqVectorMonitor is being handled correctly via _gnix_nic_progress(), then this issue can be closed.
